src/io/simulator/simulator_handle.hpp

```diff
@@ -0,0 +1,453 @@
+  explicit SimulatorHandle(SimulatorConfig config)
+      : cluster_wide_time_microseconds_(config.start_time), rng_(config.rng_seed), config_(config) {}
+
+  void IncrementServerCountAndWaitForQuiescentState(Address address) {
+    std::unique_lock<std::mutex> lock(mu_);
+    server_addresses_.insert(address);
+
+    while (true) {
+      size_t blocked_servers = blocked_on_receive_;
+
+      for (auto &[promise_key, opaque_promise] : promises_) {
+        if (opaque_promise.promise.IsAwaited()) {
+          if (server_addresses_.contains(promise_key.requester_address)) {
+            blocked_servers++;
+          }
+        }
+      }
+
+      bool all_servers_blocked = blocked_servers == server_addresses_.size();
+
+      if (all_servers_blocked) {
+        return;
+      }
+
+      cv_.wait(lock);
+    }
+  }
+
+  void TimeoutPromisesPastDeadline() {
+    uint64_t now = cluster_wide_time_microseconds_;
+
+    for (auto &[promise_key, dop] : promises_) {
+      // TODO(tyler) queue this up and drop it after its deadline
+      if (dop.deadline < now) {
+        std::cout << "timing out request" << std::endl;
+        DeadlineAndOpaquePromise dop = std::move(promises_.at(promise_key));
+        promises_.erase(promise_key);
+
+        stats_.timed_out_requests++;
+
+        dop.promise.TimeOut();
+      }
+    }
+  }
+
+  bool MaybeTickSimulator() {
+    std::unique_lock<std::mutex> lock(mu_);
+
+    size_t blocked_servers = blocked_on_receive_;
+
+    for (auto &[promise_key, opaque_promise] : promises_) {
+      if (opaque_promise.promise.IsAwaited()) {
+        if (server_addresses_.contains(promise_key.requester_address)) {
+          blocked_servers++;
+        }
+      }
+    }
+
+    if (blocked_servers < server_addresses_.size()) {
+      // we only need to advance the simulator when all
+      // servers have reached a quiescent state, blocked
+      // on their own futures or receive methods.
+      return false;
+    }
+
+    stats_.simulator_ticks++;
+
+    cv_.notify_all();
+
+    TimeoutPromisesPastDeadline();
+
+    if (in_flight_.empty()) {
+      // return early here because there are no messages to schedule
+
+      // We tick the clock forward when all servers are blocked but
+      // there are no in-flight messages to schedule delivery of.
+      std::poisson_distribution<> time_distrib(50);
+      uint64_t clock_advance = time_distrib(rng_);
+      cluster_wide_time_microseconds_ += clock_advance;
+
+      MG_ASSERT(cluster_wide_time_microseconds_ < config_.abort_time,
+                "Cluster has executed beyond its configured abort_time, and something may be failing to make progress "
+                "in an expected amount of time.");
+      return true;
+    }
+
+    if (config_.scramble_messages) {
+      // scramble messages
+      std::uniform_int_distribution<size_t> swap_distrib(0, in_flight_.size() - 1);
+      size_t swap_index = swap_distrib(rng_);
+      std::swap(in_flight_[swap_index], in_flight_.back());
```

Comment:
Have you thought about using [`std::shuffle`](https://en.cppreference.com/w/cpp/algorithm/random_shuffle)? With that you can scramble all the messages easily. Or is there a reason why to change only one message?